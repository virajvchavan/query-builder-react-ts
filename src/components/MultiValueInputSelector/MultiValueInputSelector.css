/* Most of these styles are copied from react-select's styles */
.multi-input-container {
    position: relative;
    box-sizing: border-box;
}
.multi-input-control {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    background-color: hsl(0,0%,100%);
    border-color: hsl(0,0%,80%);
    border-radius: 4px;
    border-style: solid;
    border-width: 1px;
    cursor: default;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    -webkit-box-pack: justify;
    -webkit-justify-content: space-between;
    -ms-flex-pack: justify;
    justify-content: space-between;
    min-height: 38px;
    outline: 0 !important;
    position: relative;
    -webkit-transition: all 100ms;
    transition: all 100ms;
    box-sizing: border-box;
}
.multi-select-valueContainer {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex: 1;
    -ms-flex: 1;
    flex: 1;
    -webkit-flex-wrap: wrap;
    -ms-flex-wrap: wrap;
    flex-wrap: wrap;
    padding: 2px 8px;
    -webkit-overflow-scrolling: touch;
    position: relative;
    overflow: hidden;
    box-sizing: border-box;
}
.multi-select-placeholder {
    display: none;
    color: hsl(0,0%,50%);
    margin-left: 2px;
    margin-right: 2px;
    position: absolute;
    top: 50%;
    -webkit-transform: translateY(-50%);
    -ms-transform: translateY(-50%);
    transform: translateY(-50%);
    box-sizing: border-box;
}
.multi-select-value {
    background-color: hsl(0,0%,90%);
    border-radius: 2px;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    margin: 2px;
    min-width: 0;
    box-sizing: border-box;
}
.multi-select-value-text {
    border-radius: 2px;
    color: hsl(0,0%,20%);
    font-size: 85%;
    overflow: hidden;
    padding: 3px;
    padding-left: 6px;
    text-overflow: ellipsis;
    white-space: nowrap;
    box-sizing: border-box;
}
.multi-select-value-closeBtn {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    border-radius: 2px;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    padding-left: 4px;
    padding-right: 4px;
    box-sizing: border-box;
}
.multi-select-value-closeBtn:hover {
    background-color: #FFBDAD;
    color: #DE350B;
}
.multi-select-value-closeBtn svg{
    display: inline-block;
    fill: black;
    line-height: 1;
    stroke: black;
    stroke-width: 0;
}
.multi-select-inputBox {
    width: 100%;
    margin: 2px;
    padding-bottom: 2px;
    padding-top: 2px;
    visibility: visible;
    color: hsl(0,0%,20%);
    box-sizing: border-box;
    font-size: 14px;
}
.multi-select-inputBox input {
    box-sizing: content-box;
    width: 100%;
    background: 0px center;
    border: 0px;
    font-size: inherit;
    opacity: 1;
    outline: 0px;
    padding: 0px;
    color: inherit;
}
.multi-select-BtnContainer {
    -webkit-align-items: center;
    -webkit-box-align: center;
    -ms-flex-align: center;
    align-items: center;
    -webkit-align-self: stretch;
    -ms-flex-item-align: stretch;
    align-self: stretch;
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    -webkit-flex-shrink: 0;
    -ms-flex-negative: 0;
    flex-shrink: 0;
    box-sizing: border-box;
}
.multi-select-BtnContainer .multi-select-btn {
    color: hsl(0,0%,80%);
    display: -webkit-box;
    display: -webkit-flex;
    display: -ms-flexbox;
    display: flex;
    padding: 8px;
    -webkit-transition: color 150ms;
    transition: color 150ms;
    box-sizing: border-box;
    cursor: pointer;
}
.multi-select-BtnContainer .multi-select-btn.add-all-btn {
    color: green;
    font-size: 25px;
}
.multi-select-BtnContainer .multi-select-btn svg {
    display: inline-block;
    fill: currentColor;
    line-height: 1;
    stroke: currentColor;
    stroke-width: 0;
}
.multi-select-btnSeparator {
    -webkit-align-self: stretch;
    -ms-flex-item-align: stretch;
    align-self: stretch;
    background-color: hsl(0,0%,80%);
    margin-bottom: 8px;
    margin-top: 8px;
    width: 1px;
    box-sizing: border-box;
}
