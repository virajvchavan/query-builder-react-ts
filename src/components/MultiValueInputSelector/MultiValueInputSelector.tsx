import React, { useEffect, useState } from 'react';
import './MultiValueInputSelector.css';
import { CloseIcon } from '../shared/icons';

interface Props {
    defaultValues?: Array<number>,
    onChange?: (newRhs: Array<number>) => void,
    type: "number" // can add support for other types later
}

const MultiValueInputSelector = (props: Props) => {
    const [values, setValues] = useState<Array<number>>(props.defaultValues || []);
    const [currentValue, setCurrentValue] = useState<number>();

    useEffect(() => {
        if (props.onChange) {
            props.onChange(values);
        }
    }, [values]);

    const addValue = (valueToAdd: number) => {
        setValues(values => {
            return [...values, valueToAdd];
        });
    }

    const removeValue = (valueToRemove: number) => {
        setValues(values => {
            let newValues = [...values];
            return newValues.filter(value => value !== valueToRemove);
        });
    }

    const removeAllValues = () => {
        setValues([]);
    }

    const onAddBtnClick = () => {
        if (currentValue) {
            addValue(currentValue);
        }
        if (currentValue !== null) {
            setCurrentValue(undefined);
        }
    }

    const onInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
        if (event.target.value !== null) {
            setCurrentValue(parseInt(event.target.value));
        }
    };

    const onInputKeyPress = (event: React.KeyboardEvent<HTMLInputElement>) => {
        if (event.key === "Enter") {
            onAddBtnClick();
        }
    }

    return (
        <div className="multi-input-container">
            <div className="multi-input-control">
                <div className="multi-select-valueContainer">
                    {values.map((value, index) => {
                        return <div key={index} className="multi-select-value">
                            <div className="multi-select-value-text">{value}</div>
                            <div className="multi-select-value-closeBtn" data-testid="removeBtn" data-value={value} onClick={() => { removeValue(value) }}>
                                <CloseIcon size="14" />
                            </div>
                        </div>
                    })}
                    <div className="multi-select-inputBox">
                        <div style={{display: "inline-block", width: "100%"}}>
                            <label htmlFor="multi-value-input" style={{display: "none"}}>Enter values</label>
                            <input id="multi-value-input" onChange={onInputChange} onKeyPress={onInputKeyPress} placeholder="Enter values (1-1000)" type={props.type} min="0" max="1000" value={currentValue || ""} />
                        </div>
                    </div>
                </div>
                <div className="multi-select-BtnContainer">
                    {values.length > 0 && (
                        <div className="multi-select-btn" data-testid="removeAllBtn" onClick={removeAllValues}>
                            <CloseIcon size="20" />
                        </div>
                    )}
                    <div className="multi-select-btnSeparator"></div>
                    <div className="multi-select-btn add-all-btn" onClick={onAddBtnClick}>
                        +
                    </div>
                </div>
            </div>
        </div>
    );
};

export default MultiValueInputSelector;
