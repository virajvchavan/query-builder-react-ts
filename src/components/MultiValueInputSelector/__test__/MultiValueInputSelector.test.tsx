import { cleanup, render } from "@testing-library/react";
import React from "react";
import userEvent from '@testing-library/user-event'
import MultiValueInputSelector from "../MultiValueInputSelector";

afterEach(cleanup);

test("it should have mandatory elements", () => {
    const { getByLabelText, getByText } = render(<MultiValueInputSelector type="number" />);

    expect(getByLabelText("Enter values")).not.toBeNull();
    expect(getByText("+")).not.toBeNull();
});

test("add button click should add an item", () => {
    const { container, getByLabelText, getByText } = render(<MultiValueInputSelector type="number" />);
    let input = getByLabelText("Enter values");
    let addBtn = getByText("+");
    userEvent.type(input, "102");
    userEvent.click(addBtn);
    expect(getByText("102")).not.toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(1);

    userEvent.type(input, "67");
    userEvent.click(addBtn);
    expect(getByText("67")).not.toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(2);
});

test("keypress Enter on input should add an item", () => {
    const { container, getByLabelText, getByText } = render(<MultiValueInputSelector type="number" />);
    let input = getByLabelText("Enter values");
    userEvent.type(input, "102");
    userEvent.type(input, '{enter}');
    expect(getByText("102")).not.toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(1);

    userEvent.type(input, "67");
    userEvent.type(input, '{enter}');
    expect(getByText("67")).not.toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(2);
});

test("input should be empty right after adding an element", () => {
    const { getByLabelText } = render(<MultiValueInputSelector type="number" />);
    let input = getByLabelText("Enter values");
    userEvent.type(input, "102");
    userEvent.type(input, '{enter}');
    expect(input).toHaveValue(null);
});

test("remove btn click should remove an item", () => {
    const { getByLabelText, getByText, getAllByTestId, queryByText } = render(<MultiValueInputSelector type="number" />);
    let input = getByLabelText("Enter values");
    userEvent.type(input, "102");
    userEvent.type(input, '{enter}');
    expect(getByText("102")).not.toBeNull();
    userEvent.click(getAllByTestId("removeBtn")[0]);
    expect(queryByText("102")).toBeNull();
});

test("remove-all button is shown only if there is at least one element added", () => {
    const { getByLabelText, getByText, queryByTestId, getAllByTestId } = render(<MultiValueInputSelector type="number" />);
    let input = getByLabelText("Enter values");
    let addBtn = getByText("+");

    expect(queryByTestId("removeAllBtn")).not.toBeInTheDocument();

    userEvent.type(input, "102");
    userEvent.click(addBtn);

    expect(queryByTestId("removeAllBtn")).toBeInTheDocument();

    userEvent.click(getAllByTestId("removeBtn")[0]);

    expect(queryByTestId("removeAllBtn")).not.toBeInTheDocument();
});

test("remove-all btn should remove all items", () => {
    const { container, getByLabelText, getByText, getByTestId } = render(<MultiValueInputSelector type="number" />);
    let input = getByLabelText("Enter values");
    let addBtn = getByText("+");

    userEvent.type(input, "102");
    userEvent.click(addBtn);
    userEvent.type(input, "67");
    userEvent.click(addBtn);

    userEvent.click(getByTestId("removeAllBtn"));
    expect(container.querySelectorAll(".multi-select-value").length).toBe(0);
});


test("it should call onChange method with correct params", () => {
    const onChange = jest.fn();
    const { getByLabelText, getByText } = render(<MultiValueInputSelector type="number" onChange={onChange} />);
    let input = getByLabelText("Enter values");
    let addBtn = getByText("+");

    userEvent.type(input, "102");
    userEvent.click(addBtn);
    expect(onChange).toHaveBeenLastCalledWith([102]);

    userEvent.type(input, "67");
    userEvent.click(addBtn);
    expect(onChange).toHaveBeenLastCalledWith([102, 67]);
});

test("it should follow min max validations if provided through rules prop", () => {
    const { getByLabelText, getByText, container, queryByText } = render(<MultiValueInputSelector type="number" rules={["range:1-1000"]} />);
    let input = getByLabelText("Enter values");
    let addBtn = getByText("+");
    userEvent.type(input, "102");
    userEvent.click(addBtn);
    expect(getByText("102")).not.toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(1);

    // add a number greater than max
    userEvent.type(input, "1200");
    userEvent.click(addBtn);
    expect(queryByText("1200")).toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(1);

    // value should not be cleared
    expect(input).toHaveValue(1200);

    // clear the input
    userEvent.clear(input);

    userEvent.type(input, "-24");
    userEvent.click(addBtn);
    expect(queryByText("-24")).toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(1);
});

test("it should match snapshot when type is number; for first render", () => {
    const { asFragment } = render(<MultiValueInputSelector type="number"  />)
    expect(asFragment()).toMatchSnapshot();
});

test("it should match snapshot when type is number & range is specified; for first render", () => {
    const { asFragment } = render(<MultiValueInputSelector type="number" rules={["range:1-1000"]} />)
    expect(asFragment()).toMatchSnapshot();
});

test("it should match snapshot when an element is added", () => {
    const { asFragment, container, getByLabelText, getByText } = render(<MultiValueInputSelector type="number" />);
    let input = getByLabelText("Enter values");
    let addBtn = getByText("+");
    userEvent.type(input, "102");
    userEvent.click(addBtn);
    expect(getByText("102")).not.toBeNull();
    expect(container.querySelectorAll(".multi-select-value").length).toBe(1);
    expect(asFragment()).toMatchSnapshot();
});
