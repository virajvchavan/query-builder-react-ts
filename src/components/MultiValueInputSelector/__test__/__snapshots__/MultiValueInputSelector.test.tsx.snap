// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it should match snapshot when an element is added 1`] = `
<DocumentFragment>
  <div
    class="multi-input-container"
  >
    <div
      class="multi-input-control"
    >
      <div
        class="multi-select-valueContainer"
      >
        <div
          class="multi-select-value"
        >
          <div
            class="multi-select-value-text"
          >
            102
          </div>
          <div
            class="multi-select-value-closeBtn"
            data-testid="removeBtn"
            data-value="102"
          >
            <svg
              aria-hidden="true"
              focusable="false"
              height="14"
              viewBox="0 0 20 20"
              width="14"
            >
              <path
                d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
              />
            </svg>
          </div>
        </div>
        <div
          class="multi-select-inputBox"
        >
          <div
            style="display: inline-block; width: 100%;"
          >
            <label
              for="multi-value-input"
              style="display: none;"
            >
              Enter values
            </label>
            <input
              id="multi-value-input"
              max="1000"
              min="0"
              placeholder="Enter values (0-1000)"
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="multi-select-BtnContainer"
      >
        <div
          class="multi-select-btn"
          data-testid="removeAllBtn"
        >
          <svg
            aria-hidden="true"
            focusable="false"
            height="20"
            viewBox="0 0 20 20"
            width="20"
          >
            <path
              d="M14.348 14.849c-0.469 0.469-1.229 0.469-1.697 0l-2.651-3.030-2.651 3.029c-0.469 0.469-1.229 0.469-1.697 0-0.469-0.469-0.469-1.229 0-1.697l2.758-3.15-2.759-3.152c-0.469-0.469-0.469-1.228 0-1.697s1.228-0.469 1.697 0l2.652 3.031 2.651-3.031c0.469-0.469 1.228-0.469 1.697 0s0.469 1.229 0 1.697l-2.758 3.152 2.758 3.15c0.469 0.469 0.469 1.229 0 1.698z"
            />
          </svg>
        </div>
        <div
          class="multi-select-btnSeparator"
        />
        <div
          class="multi-select-btn add-all-btn"
        >
          +
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`it should match snapshot when type is number & range is specified; for first render 1`] = `
<DocumentFragment>
  <div
    class="multi-input-container"
  >
    <div
      class="multi-input-control"
    >
      <div
        class="multi-select-valueContainer"
      >
        <div
          class="multi-select-inputBox"
        >
          <div
            style="display: inline-block; width: 100%;"
          >
            <label
              for="multi-value-input"
              style="display: none;"
            >
              Enter values
            </label>
            <input
              id="multi-value-input"
              max="1000"
              min="1"
              placeholder="Enter values (1-1000)"
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="multi-select-BtnContainer"
      >
        <div
          class="multi-select-btnSeparator"
        />
        <div
          class="multi-select-btn add-all-btn"
        >
          +
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`it should match snapshot when type is number; for first render 1`] = `
<DocumentFragment>
  <div
    class="multi-input-container"
  >
    <div
      class="multi-input-control"
    >
      <div
        class="multi-select-valueContainer"
      >
        <div
          class="multi-select-inputBox"
        >
          <div
            style="display: inline-block; width: 100%;"
          >
            <label
              for="multi-value-input"
              style="display: none;"
            >
              Enter values
            </label>
            <input
              id="multi-value-input"
              max="1000"
              min="0"
              placeholder="Enter values (0-1000)"
              type="number"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        class="multi-select-BtnContainer"
      >
        <div
          class="multi-select-btnSeparator"
        />
        <div
          class="multi-select-btn add-all-btn"
        >
          +
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
